### PR TITLE
Add support for stopping connectors

### PR DESCRIPTION
**Pull request by mimaison**
### Type of change

- Enhancement / new feature

### Description

This implements [Proposal 54](https://github.com/strimzi/proposals/blob/main/054-stopping-kafka-connect-connectors.md)

It adds a new field for connectors, state, that can be set to running (the default), paused or stopped. The existing pause field is now deprecated and if both fields are set, state takes precedence.

The proposal also mentions handling Kafka Connect < 3.5 and downgrade stop to pause in this case. From what I understand the next version of Strimzi will only support Kafka >= 3.5 so the `PUT /connector/<CONN>/stop` should be available. For that reason I've not added logic to handle Kafka Connect < 3.5.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 1]
```diff
@@ -475,27 +476,64 @@ protected Future<Map<String, Object>> createOrUpdateConnector(Reconciliation rec
...
-                return apiClient.resume(reconciliation, host, port, connectorName);
-            } else {
-                return Future.succeededFuture();
+            ConnectorState desiredState = connectorSpec.getState();
+            @SuppressWarnings("deprecation")
+            Boolean shouldPause = connectorSpec.getPause();
+            ConnectorState effectiveState = desiredState != null ? desiredState :
+                    Boolean.TRUE.equals(shouldPause) ? ConnectorState.PAUSED : ConnectorState.RUNNING;
+            if (desiredState != null && shouldPause != null) {
+                String message = "Both pause and state are set. Since pause is deprecated, state takes precedence " +
```
Comment (mimaison) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:495]: Should I use `$.connector.` like in the message above, or is this fine?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 2]
```diff
@@ -707,11 +745,11 @@ protected boolean hasRestartAnnotation(CustomResource resource, String connector
     }
 
     /**
-     * Return the ID of the connector task to be restarted if the provided KafkaConnector resource instance has the strimzio.io/restart-task annotation
+     * Return the ID of the connector task to be restarted if the provided KafkaConnector resource instance has the strimzi.io/restart-task annotation
```
Comment (mimaison) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:748]: These 2 changes are unrelated. Let me know if you prefer to fix them in a separate PR.

**Review by scholzj — COMMENTED** [review]
I left some nits. But LGTM otherwise. Also, it looks like the CI failure is related to this:
```
[ERROR] Failures: 
[ERROR] io.strimzi.operator.cluster.operator.assembly.KafkaConnectApiIT.test(VertxTestContext)
[ERROR]   Run 1: KafkaConnectApiIT.test(VertxTestContext) io.vertx.core.impl.NoStackTraceThrowable: Unexpected status code 204 for GET request to localhost:37969/connectors/test/stop
[ERROR]   Run 2: KafkaConnectApiIT.test(VertxTestContext) io.vertx.core.impl.NoStackTraceThrowable: Unexpected status code 204 for GET request to localhost:46515/connectors/test/stop
[ERROR]   Run 3: KafkaConnectApiIT.test(VertxTestContext) io.vertx.core.impl.NoStackTraceThrowable: Unexpected status code 204 for GET request to localhost:34987/connectors/test/stop
```

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java (hunk 1 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:495]: No, if you want, you can do `.spec.state` and `.spec.pause`. But not the `$`.

File: api/src/main/java/io/strimzi/api/kafka/model/ConnectorState.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+public enum ConnectorState {
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/ConnectorState.java:10]: I would add even just one line of Javadoc here.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java (hunk 1 above)
Comment (mimaison) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:495]: Ok, I'll leave it as is

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 4]
```diff
@@ -475,27 +476,64 @@ protected Future<Map<String, Object>> createOrUpdateConnector(Reconciliation rec
...
+                String message = "Both pause and state are set. Since pause is deprecated, state takes precedence " +
+                        "so the connector will be " + effectiveState.toValue();
+                LOGGER.warnCr(reconciliation, message);
+                conditions.add(StatusUtils.buildWarningCondition("UpdateState", message));
+            }
+            Future<Void> future = Future.succeededFuture();
+            switch (state) {
+                case "RUNNING" -> {
+                    if (effectiveState == ConnectorState.PAUSED) {
+                        LOGGER.debugCr(reconciliation, "Pausing connector {}", connectorName);
```
[the remark this answers is not in the document]
Comment (mimaison) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java]: I changed them to INFO level

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 5]
```diff
@@ -475,27 +476,64 @@ protected Future<Map<String, Object>> createOrUpdateConnector(Reconciliation rec
...
-                return apiClient.pause(reconciliation, host, port, connectorName);
-            } else if ("PAUSED".equals(state) && !shouldPause) {
-                LOGGER.debugCr(reconciliation, "Resuming connector {}", connectorName);
-                return apiClient.resume(reconciliation, host, port, connectorName);
-            } else {
-                return Future.succeededFuture();
+            ConnectorState desiredState = connectorSpec.getState();
+            @SuppressWarnings("deprecation")
+            Boolean shouldPause = connectorSpec.getPause();
+            ConnectorState effectiveState = desiredState != null ? desiredState :
```
[the remark this answers is not in the document]
Comment (mimaison) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java]: I renamed the variable to `targetState`

File: api/src/main/java/io/strimzi/api/kafka/model/ConnectorState.java (hunk 3 above)
Comment (mimaison) [reply api/src/main/java/io/strimzi/api/kafka/model/ConnectorState.java:13]: I added a short javadoc, let me know if it's not what you had in mind

**Comment by mimaison** [comment]
In `KafkaConnectApiImpl`, in most calls to the Connect REST API, in case of errors, we seem to wrap the error into a `ConnectRestException` when failing the future. For example: https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiImpl.java#L92

This was not done in `[pauseResume()](https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiImpl.java#L302)` (now renamed to `updateState()`). Is there a specific reason or should we make it consistent?

**Comment by ppatierno** [comment]
> Is there a specific reason or should we make it consistent?

@mimaison I don't know I am afraid. I noticed there are a couple of other places where the wrapping doesn't happen (i.e. one in `delete` and one in `list`).

**Comment by scholzj** [comment]
@tombentley wrote this if I remember it correctly. Maybe he would have an idea? Obviously, it is a long time ago.

**Review by scholzj — APPROVED** [review]
LGTM. Thanks.

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks!

**Comment by mimaison** [comment]
For the `Future` `String`/`ConnectRestException` inconsistency, I propose fixing that in a follow up PR.

**Comment by mimaison** [comment]
@scholzj Done!
